src/types/redis_bitmap.cc

```diff
@@ -538,6 +545,281 @@ rocksdb::Status Bitmap::BitOp(BitOpFlags op_flag, const std::string &op_name, co
+    if (!view_status.ok()) {
+      return view_status;
+    }
+
+    uint32_t first_bit_segment_offset = offset % kBitmapSegmentBits;
+    if (enc.IsSigned()) {
+      *res = GetSignedBitfield(view, first_bit_segment_offset, enc.Bits());
+    } else {
+      *res = GetUnsignedBitfield(view, first_bit_segment_offset, enc.Bits());
+    }
+
+    return rocksdb::Status::OK();
+  };
+
+  if constexpr (ReadOnly) {
+    std::string ret;
+    for (BitfieldOperation op : ops) {
+      if (op.type != BitfieldOperation::Type::kGet) {
+        return rocksdb::Status::InvalidArgument("Write bitfield in read-only mode.");
+      }
+
+      uint64_t unsigned_old_value = 0;
+      s = get(op.offset, op.encoding, &unsigned_old_value);
+      if (!s.ok()) {
+        return s;
+      }
+
+      if (op.encoding.IsSigned()) {
+        ret = redis::Integer(CastToSignedWithoutBitChanges(unsigned_old_value));
+      } else {
+        ret = redis::Integer(unsigned_old_value);
+      }
+
+      rets->push_back(std::move(ret));
+    }
+    return rocksdb::Status::OK();
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  bitfieldWriteAheadLog(batch, ops);
+
+  // set bitfield value among segments
+  auto set = [&](uint32_t offset, BitfieldEncoding enc, BitfieldOverflowBehavior overflow, uint64_t value) {
+    BitmapSegmentsView view;
+    s = init_view(offset, enc.Bits(), &view);
+    if (!s.ok()) {
+      return s;
+    }
+
+    uint32_t first_bit_segment_offset = offset % kBitmapSegmentBits;
+    SetBitfield(view, first_bit_segment_offset, enc.Bits(), value);
+
+    for (auto segment = view.SegmentBegin(); segment != view.SegmentEnd(); ++segment) {
+      batch->Put((*segment)->first, (*segment)->second);
+    }
+
+    uint64_t segments_cnt = view.SegmentsCount();
+    uint64_t used_size = (offset / 8) + (segments_cnt - 1) * kBitmapSegmentBytes + view.LastSegmentSize();
+    if (metadata.size < used_size) {
+      metadata.size = used_size;
+      std::string bytes;
+      metadata.Encode(&bytes);
+      batch->Put(metadata_cf_handle_, ns_key, bytes);
+    }
+
+    return rocksdb::Status::OK();
+  };
+
+  std::string ret;
+  for (BitfieldOperation op : ops) {
+    uint64_t unsigned_old_value = 0;
+    s = get(op.offset, op.encoding, &unsigned_old_value);
+    if (!s.ok()) {
+      return s;
+    }
+
+    uint64_t unsigned_new_value = 0;
+    if (BitfieldOp(op, unsigned_old_value, &unsigned_new_value)) {
+      if (op.type != BitfieldOperation::Type::kGet) {
+        s = set(op.offset, op.encoding, op.overflow, unsigned_new_value);
+        if (!s.ok()) {
+          return s;
+        }
+      }
+
+      if (op.type == BitfieldOperation::Type::kSet) {
+        unsigned_new_value = unsigned_old_value;
+      }
+
+      if (op.encoding.IsSigned()) {
+        ret = redis::Integer(CastToSignedWithoutBitChanges(unsigned_new_value));
+      } else {
+        ret = redis::Integer(unsigned_new_value);
+      }
+    } else {
+      ret = redis::NilString();
+    }
```

Comment:
I think it would be better to return `vector<optional<integer>>` here rather than vector of RESP strings here, in case that we can reuse the result.

Reply:
Should I use std::variant<nullptr_t, uint64_t, int64_t> or define a new type to handle this? There are 3 of type could store in result list (null, unsigned and signed). These two way above can store result in 16 bytes. But it doesn't seem intuitive enough.

Reply:
Maybe it can be `std::optional<struct { uint64_t v; bool signed; }>`, if you are willing to change it.